First 4me connector version

den-sergeyev-workato opened the pull request:
First 4me connector version

Commit:
4me initial commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2798 @@
 {
-  title: 'Template Partner Repo',
+  title: '4me',
+  secure_tunnel: false,
```
Definition that doesn't do anything, probably we can omit it?

klaasvandeweerdt replied:
Yes, it can be omitted.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2798 @@
+      {
+        name: 'account',
+        label: '4me Account',
+        optional: false,
```
`false` is default value for optional

klaasvandeweerdt replied:
When using the online code editor and debug console I have to explicitly specify it, otherwise, the fields are not required in the debug console

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2798 @@
+      },
+      {
+        name: 'auth_method',
+        control_type: 'select',
```
Hint can be added.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2798 @@
+        }
+      }",
+      nil,
+      connection['account']
```
Why we cannot get account from connection inside the function? Hence we have can reduce number of params for `run_gql` method.

klaasvandeweerdt replied:
The default account value is the one from the connection, but in every action, I can specify another account. The authorization can grant me access to multiple 4me accounts, which is why the param is required in the `run_gql`.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2798 @@
+
+          apply: lambda do |connection|
+            if current_url.include?('https://graphql.')
+              token_type = connection['token_type'] || 'Bearer'
```
Did we forget to define `token_type` or remove it here?

klaasvandeweerdt replied:
It can be removed, the token type is always `Bearer`

den-sergeyev-workato commented:
@klaasvandeweerdt Can you, please, fix rubocop?
```
63 files inspected, 571 offenses detected, 458 offenses autocorrectable
```

den-sergeyev-workato commented:
@klaasvandeweerdt We have 28 pending specs, are you going to implement them?

klaasvandeweerdt commented:
@den-sergeyev-workato, I will remove the pending specs as they are not required, still part of the auto-generate and fix rubocop.

klaasvandeweerdt commented:
@den-sergeyev-workato, I updated the code based on the comments, remove the pending specs, and fixed all rubocop offenses. Also ran a bunch of tests via the online code editor and debug console as not all methods are supported.

Commit:
Added recommendation, fixed rubocop and removed unused pending specs

den-sergeyev-workato commented:
> I updated the code based on the comments, remove the pending specs, and fixed all rubocop offenses. Also ran a bunch of tests via the online code editor and debug console as not all methods are supported.

@bennettgo I believe we need to add our rubocop config to the template we share with partners.

den-sergeyev-workato commented:
@klaasvandeweerdt Can you, please, add logo.png to the repository?

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+                                      client_secret: connection['client_secret']
+                                    })
+            request = request.payload(payload)
+            request.request_format_www_form_urlencoded
```
Per our [style guides](https://docs.workato.com/developing-connectors/sdk/sdk-reference/connection/authorization.html#acquire) we recommend to return hash explicitly.

klaasvandeweerdt replied:
Implemented, will be in next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
   actions: {
+    query: {
+      title: 'Query records',
+      subtitle: 'Retrieve one or more records.',
```
For subtitles in generic methods it is recommend to add app name and example
```
Retrieve one or more records, e.g. user, in 4me
```

The same logic should be applied to all subtitles for all actions

klaasvandeweerdt replied:
Implemented, will be in next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+        label = name&.labelize&.downcase
+        "Query <span class='provider'>" \
+          "#{label || 'records'}</span> via " \
+          "<span class='provider'>GraphQL</span>"
```
Please, add application name and highlight it with <span class='provider'></span>
```
in <span class='provider'>4me</span>
```

The same logic should be applied to all descriptions for all actions

klaasvandeweerdt replied:
Implemented, will be in next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+          "<span class='provider'>GraphQL</span>"
+      end,
+      input_fields: lambda do |object_definitions|
+        object_definitions['query_input']
```
How large can be result list here? Can you provide, please, provide example of largest input possible here?

klaasvandeweerdt replied:
Implemented, will be in next commit

klaasvandeweerdt replied:
There are about 80 root connections, each connection contains on average 15 properties and 5 other connections. Our API support connection till 13 levels deep.
This is an example of a 13-connections deep query, with a selection of the most commonly use fields.

```
{
  "account": "wdc",
  "object": "people{}",
  "argument_first": "10",
  "argument_view": "all",
  "nested_fields": "nodes",
  "primitive_fields": "totalCount",
  "field_nodes": {
    "primitive_fields": "authenticationID,id,primaryEmail,supportID,timeFormat24h,timeZone,updatedAt,vip",
    "nested_fields": "teams,site,organization,manager",
    "field_teams": {
      "argument_first": "10",
      "primitive_fields": "totalCount",
      "nested_fields": "nodes",
      "field_nodes": {
        "primitive_fields": "autoAssign,createdAt,disabled,id,inboundEmailLocalPart,name,pictureUri,remarks,source,sourceID,timeZone,updatedAt",
        "nested_fields": "members",
        "field_members": {
          "argument_first": "10",
          "nested_fields": "nodes",
          "primitive_fields": "",
          "field_nodes": {
            "primitive_fields": "id,name,primaryEmail,employeeID",
            "nested_fields": "configurationItems",
            "field_configurationItems": {
              "argument_first": "10",
              "primitive_fields": "totalCount",
              "nested_fields": "nodes",
              "field_nodes": {
                "nested_fields": "product",
                "primitive_fields": "name,serialNr,status",
                "field_product": {
                  "primitive_fields": "id,name,model",
                  "nested_fields": "service",
                  "field_service": {
                    "primitive_fields": "id,name",
                    "nested_fields": "supportTeam,account",
                    "field_account": {
                      "primitive_fields": "disabled,id,name"
                    },
                    "field_supportTeam": {
                      "primitive_fields": "name,timeZone"
                    }
                  }
                }
              }
            }
          }
        }
      }
    },
    "field_site": {
      "primitive_fields": "createdAt,disabled,id,name,pictureUri,remarks,source,sourceID,timeZone,updatedAt",
      "nested_fields": "people",
      "field_people": {
        "argument_first": "10",
        "nested_fields": "nodes",
        "field_nodes": {
          "primitive_fields": "id,name,primaryEmail",
          "nested_fields": "skillPools",
          "field_skillPools": {
            "argument_first": "10",
            "primitive_fields": "totalCount",
            "nested_fields": "nodes",
            "field_nodes": {
              "primitive_fields": "id,name,disabled"
            }
          }
        }
      }
    }
  }
}
```

den-sergeyev-workato replied:
Is it possible to use [Summarize input](https://docs.workato.com/developing-connectors/sdk/sdk-reference/actions.html#summarize-input) here?

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+        call('action_execute', connection, input, eis, eos, 'query')
+      end,
+      output_fields: lambda do |object_definitions|
+        object_definitions['query_output']
```
How large can be list here?

klaasvandeweerdt replied:
The default, and maximum, the number of items per connection is 100, and the maximum number of items that can be returned in a single request with multiple connections is limited to 500,000. When querying nested data, such as 100 people with 100 skill pools and 100 members each, the total number of items (100 x 100 x 100) exceeds the maximum of 500,000.

den-sergeyev-workato replied:
In worst case we should use [summarize-output)](https://docs.workato.com/developing-connectors/sdk/sdk-reference/actions.html#summarize-output) then.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+        call('action_execute', connection, input, eis, eos, 'query')
+      end,
+      output_fields: lambda do |object_definitions|
+        object_definitions['query_output']
```
Is it possible to provide [sample output](https://docs.workato.com/developing-connectors/sdk/sdk-reference/actions.html#sample-output)?

klaasvandeweerdt replied:
@den-sergeyev-workato, this will be a challenge. We have about 420 data entities, so hard-coding them is not an option. I could make it dynamic but that will require a GraphQL query for each field or sub-query selection.  It takes me 20 to 25 click to build my query, this would also mean 20 to 25 request, and we have rate-limiting in place.
I could add the most commonly used data model with a basic sample but that will be a few hundred lines of code.
I will look into it and see what I can do, any suggestions?

den-sergeyev-workato replied:
I see the complexity.
It is not mandatory for now, so we can make as separate improvement if find handy way

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+    upload_attachment: {
+      title: 'Upload attachment',
+      description: 'Upload a file',
+
```
Can we add help section here?

klaasvandeweerdt replied:
Implemented, will be in next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+        body: 'Use this action to create, delete or update records.',
+        learn_more_url: 'https://developer.4me.com/graphql/',
+        learn_more_text: 'Learn more'
+      },
```
Does make sense to label it as [batch](https://docs.workato.com/developing-connectors/sdk/sdk-reference/actions.html#batch) action?

klaasvandeweerdt replied:
This is, currently, always per item. I change the body to `Use this action to create, delete or update a record.`

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+      "{#{fields&.join(' ')}}" if fields.present?
+    end,
+
+    build_query_field_list_: lambda do |input, eis, eos|
```
Do we still need it?
There are both `build_query_field_list_` and `build_query_field_list`

klaasvandeweerdt replied:
Two methods I copied from the GraphQL connector, `build_query_field_list_` is not used, removed it.
will be in next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
-  object_definitions: {
+  triggers: {
+    new_event: {
+      title: 'Webhook event',
```
Assuming this is generic trigger for all types of records, we have recommendation to describe it as record 
```
New/updated record
```
However if this doesn't fit your semantics I'm open to discuss it.

klaasvandeweerdt replied:
@den-sergeyev-workato, `New/updated record` does not cover all functionality. From within 4me, we can invoke automation rules. We have a webhook event name `automation_rule` which contains a custom build payload and it is unrelated to a record.
Would `New/updated record, and automation rule events` make sense?

klaasvandeweerdt replied:
@den-sergeyev-workato, had a chat with our team about it. `Webhook event` is a known and commonly used term in 4me development and integrations. In favor of keeping `Webhook event` with `Triggers when a selected 4me object, e.g person, is created/updated, or on an automation rule notification` as the subtitle.

den-sergeyev-workato replied:
I believe this is too generic trigger, I'll double check wether we can skip this requirement in our checklist.

den-sergeyev-workato replied:
@klaasvandeweerdt got approve to skip naming requirement for this trigger.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+  triggers: {
+    new_event: {
+      title: 'Webhook event',
+      subtitle: 'Triggers when a Webhook event is received from 4me',
```
For generic triggers, add a common object name as example following the pattern:
```
Triggers when selected 4me object, e.g person, is created/updated
```

klaasvandeweerdt replied:
Similar to my previous note about the webhook title. Changed it to
```
Triggers when a selected 4me object, e.g person, is created/updated, or on an automation rule notification
```

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+        ['Time entry create', 'time_entry.create'],
+        ['Time entry delete', 'time_entry.delete'],
+        ['Time entry update', 'time_entry.update'],
+        # ['Webhook verify','webhook.verify'],
```
Do we still need it?

klaasvandeweerdt replied:
No, removed it, will be in next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+      subtitle: 'Triggers when a Webhook event is received from 4me',
+
+      description: lambda do |input, picklist_label|
+        "New <span class='provider'>webhook</span> in <span class='provider'>4me</span>"
```
Is it possible to make it dynamic based on `event_selection`?

klaasvandeweerdt replied:
@den-sergeyev-workato, unfortunately not. This is something I added to our backlog but will not be released soon.
The payload also depends on the `event_selection`.

den-sergeyev-workato replied:
Let me check if this mandatory requirement can be skipped.

den-sergeyev-workato replied:
@klaasvandeweerdt got approve to allow to postpone fixing this for this description.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+      subtitle: 'Triggers when a Webhook event is received from 4me',
+
+      description: lambda do |input, picklist_label|
+        "New <span class='provider'>webhook</span> in <span class='provider'>4me</span>"
```
Can we follow the recommended nomenclature “<Adjective> <object> in <app>”?
e.g 
```
New/updated record in 4me
```

klaasvandeweerdt replied:
Yes, changed it as part of the previous conversation
```
Triggers when a selected 4me object, e.g person, is created/updated, or on an automation rule notification
```

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+# frozen_string_literal: true
 
+{
   connection: {
```
I don't see title anymore.

klaasvandeweerdt replied:
Sorry, added it again. Will be in next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+
+      output_fields: lambda do |object_definitions|
+        object_definitions['file_upload_output']
+      end
```
Can we define `sample_output`?

klaasvandeweerdt replied:
Added, will be in next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+      end,
+
+      dedup: lambda do |record|
+        "#{record['account_id']}_#{record['event']}_#{record['object_id']}_#{Time.now.to_f}"
```
Technically we have all records unique due to `Time.now.to_f`, do we want to achieve this or we do need some deduplication?

If we need deduplication probably we can introduce some field that corresponds to some updated at timestamp?

klaasvandeweerdt replied:
We don't have any data available in the webhook payload to implement deduplication. In 4me there are mechanisms in place to avoid duplication.

den-sergeyev-workato replied:
Got it, so technically you don't need any deduplication here and just wanted to make unique signature for all events.

klaasvandeweerdt replied:
Yes

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+          <b>Webhook endpoint URL</b>
+          <b class="tips__highlight">#{webhook_base_url}</b>
+          More information on how to use 4me automation rules and webhooks can be found on the <a href="https://developer.4me.com//v1/workato_connector/" target="_blank">4me developer pages</a>.
+        HTML
```
Can we add `learn_more_url` and `learn_more_text`?

klaasvandeweerdt replied:
@den-sergeyev-workato 
The link is embedded in the HTML message.
```
More information on how to use 4me automation rules and webhooks can be found on the <a href="https://developer.4me.com//v1/workato_connector/" target="_blank">4me developer pages</a>.
```
This page is not available yet and will be published if the connector gets released

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+      end,
+
+      output_fields: lambda do |object_definitions|
+        object_definitions['webhook_event']
```
How many records we can have here? Should we think about [summarize](https://docs.workato.com/developing-connectors/sdk/sdk-reference/triggers.html#summarize-output)

klaasvandeweerdt replied:
This will always be a single record. The record contains the generic data, available in all webhooks and the payload property contains the basic data, or custom-defined data if an `automation_rule` event is used.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+        # Operation names
+        operation_names = documents.map do |doc|
+          doc[:operation_name]
+        end&.compact
```
result is never nil here, so safe navigation is useless.

NIT: it could be rewritten like this
```
  documents = [] 

  if  query.present?
       .....
     documents = parsed_query[:documents]
  end

       ....
  # Operation names
   operation_names = documents.map{ |doc| doc[:operation_name] }.compact
```

klaasvandeweerdt replied:
Updated, will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+      end,
+
+      execute: lambda do |connection, input|
+        post(input['webhook_verify'])
```
Does it use input as relative path of absolute one?

klaasvandeweerdt replied:
Absolute.

den-sergeyev-workato replied:
Is it possible to use relative path here?

klaasvandeweerdt replied:
No, the webhook callback is always absolute. It has no relation with the GraphQL URL.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+                   .payload(
+                     'Content-Type': input['content_type'],
+                     'acl': provider_parameters['acl'],
+                     'key': provider_parameters['key'].sub('${filename}', input['file_name']),
```
Do we need to handle input cases when input is `nil` as here? Or input couldn't be `nil` but may miss account key?
```
 account = input&.[]('account')
```

klaasvandeweerdt replied:
No, it was for missing a key. This was part of a temporary workaround, the `account` input field is now required for every action (except webhook verification).

klaasvandeweerdt replied:
Changed it to input['account'] as it can never be nil.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+          nil,
+          account
+        )['attachmentStorage']
+        provider_parameters = attachment_storage['providerParameters']
```
Is there any chance that `attachment_storage` will be nil here and we get undefined method on nil? Do we need to handle it and show meaningful user error?

klaasvandeweerdt replied:
No, it can never be nil. In case of an issue retrieving the `run_gql` lambda processed the error.

den-sergeyev-workato replied:
It will be `nil` if call(run_gql, ...) doesn't fail but has no `attachmentStorage` key in response. 
As this is request to external system I believe connector cannot guarantee that it won't have error due to missing key?

klaasvandeweerdt replied:
@den-sergeyev-workato, our API returns an error if it can not retrieve the `attachmentStorage`, it will never return a response with a missing key.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+        documents = parsed_query&.[](:documents)
+        if documents.present? && documents.length > 1
+          operation_name = input['operation_name']
+          operation = parsed_query&.[](:documents)&.find do |doc|
```
Can substitute
```
parsed_query&.[](:documents)&.
```
with
```
documents.
```

klaasvandeweerdt replied:
Updated, will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+          operation = parsed_query&.[](:documents)&.find do |doc|
+            doc[:operation_name] == operation_name
+          end
+        elsif documents.present? && documents.length == 1
```
NIT: probably easier
```
if documents.present?
  if documents.length > 1
      operation = documents.find { |doc| doc[:operation_name] ==  input['operation_name'] }
   else 
      operation = documents.first
  end
end
```

Why we cannot use find for both of them? For case with one document we don't care about `operation name`?

will it work?
```
if documents.present?
  operation = documents.find { |doc| doc[:operation_name] ==  input['operation_name'] }
end
```

klaasvandeweerdt replied:
Yes, updated, will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+        variable_name = field_name[9..-1]
+
+        variable_value = input&.[]("variable_#{variable_name}")
+        variables[variable_name] = variable_value
```
Are we ok to create key `variable_name` if when `variable_value` is nil?
```
h = {}
h[:s] = nil

 h.key?(:s)
=> true

 h.keys
=> [:s]



```

klaasvandeweerdt replied:
Updated, will be in the next commit.

klaasvandeweerdt replied:
Updated it again back to the always create a variable also when `nil`. I added the following help message:
`Every variable, including those that are empty or null, will be submitted.`
Will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+      call(
+        'run_gql',
+        connection,
+        input['query'],
```
will this fail if input `nil`?

klaasvandeweerdt replied:
Input can never be nil. In `Query records` and `Mutate records` the `select` is mandatory, and in `Custom action` the `plain-text-area` is as well.

den-sergeyev-workato replied:
I asked because code pretend to handle `nil` input several lines above
```
variable_value = input&.[]("variable_#{variable_name}")
```

klaasvandeweerdt replied:
Updated, will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+      )
+    end,
+
+    parse_graphql_type: lambda do |connection, messages, type_string|
```
Why do we need connection param here?

klaasvandeweerdt replied:
Copy/paste form the GraphQL connector will be removed in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+      end&.compact
+    end,
+
+    parse_graphql_fields: lambda do |connection, report_problem, fields_string|
```
Why do we need connection here?

klaasvandeweerdt replied:
Copy/paste form the GraphQL connector will be removed in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+      end
+    end,
+
+    parse_graphql_variables: lambda do |connection, messages, variables_string|
```
I would name it `report_problem` for consistency

klaasvandeweerdt replied:
Updated, will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+      end&.flatten(1)&.compact
+    end,
+
+    parse_graphql: lambda do |connection, report_problem, query|
```
In the end we don't need connection here is well?

klaasvandeweerdt replied:
Copy/paste form the GraphQL connector will be removed in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+      groups = query&.scan(/(?<group>[^{}]*(?<fields>{(?<inner>[^{}]|\g<fields>)*})?)/m)
+
+      # parse documents
+      # TODO: are these called 'operations' or 'documents'?
```
Do we need to resolve this TODO?

klaasvandeweerdt replied:
No, removed the comments

Commit:
Add logo

Commit:
Added JWT for webhooks

Commit:
Added tape_library

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+          value = response[key]
+          next if value.nil? || message.present?
+
+          message = call('get_error_message', connection, value)
```
For multiple errors in hash it will return only first one. Is it expected or we never have multiple keys with errors?

klaasvandeweerdt replied:
For a hash yes, in the case of an array, all messages are concatenated with `\n`, and `errors[]` is the GraphQL standard.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+      }.compact
+
+      request = post('')
+      request.headers('x-4me-Account': account)
```
Does it work fine if account is nil here? or we nee to have connection["account"] as fallback?

klaasvandeweerdt replied:
`account` can never be nil, although added `account ||= connection['account']`.
Will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+          {
+            type: :string,
+            control_type: :select,
+            pick_list: pick_list,
```
Is fine if pick_list is `nil` here or we need fallback?

klaasvandeweerdt replied:
There is no fallback, the `enumValues` are retrieved from the GraphQL schema.

den-sergeyev-workato replied:
But here you handle nil value
```
 pick_list = enum_type&.[]('enumValues')&.map do |e|
```

klaasvandeweerdt replied:
Updated, will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+          {
+            type: :string,
+            control_type: :text,
+            pick_list: nil,
```
Do I understand correctly that this is intentional remove of `pick_list` and `optional`?

klaasvandeweerdt replied:
This was part of the GraphQL connector. Yes, in case the input becomes a string instead of a list it clears the `pick_list` values as there is no need for that list on a `text` control. Same for `optional`, if some decided to not use the list they can send a `nil` value.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+          connection,
+          report_problem,
+          type['ofType'],
+          nesting_level
```
Don't we need to increment nesting level when recursively call it?

klaasvandeweerdt replied:
Not for `NOT_NULL`, it can only occur once for a type

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+      field
+    end,
+
+    apply_field_name_and_label: lambda do |connection, field, name_prefix, name|
```
Why do you need connection here?

klaasvandeweerdt replied:
Copy/paste form the GraphQL connector will be removed in the next commit.

den-sergeyev-workato replied:
Connection is still present?

klaasvandeweerdt replied:
A mistake on my end. Updated it and will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+        description = description.gsub(/\R/, '<br>')
+        description = description.gsub('\\n', '<br>')
+
+        allow_multi_paragraph_hint = false # call('get_allow_multi_paragraph_hint', connection)
```
Do we need to uncomment it?

klaasvandeweerdt replied:
The multiline logic was copied from the GraphQL connector. There is no need to support multi `get_allow_multi_paragraph_hint`. I removed all code for `allow_multi_paragraph_hint`. 
It will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+        allow_multi_paragraph_hint = false # call('get_allow_multi_paragraph_hint', connection)
+        first_break_index_one = description.index('<br>')
+        first_break_index_two = description.index('<br/>')
+        if !first_break_index_one.nil? && !first_break_index_two.nil?
```
NIT: Probably easier

```
   if !first_break_index_one.nil? && !first_break_index_two.nil?
          first_break_index = [first_break_index_one, first_break_index_two].min
        elsif first_break_index_one.nil? && !first_break_index_two.nil?
          first_break_index = first_break_index_two
        elsif !first_break_index_one.nil? && first_break_index_two.nil?
          first_break_index = first_break_index_one
        end
```

Can be converted to
```
 first_break_index = [ first_break_index_one, first_break_index_two].compact.min
```

klaasvandeweerdt replied:
The multiline logic was copied from the GraphQL connector. There is no need to support multi `get_allow_multi_paragraph_hint`. I removed all code for `allow_multi_paragraph_hint`. 
It will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+        elsif !first_break_index_one.nil? && first_break_index_two.nil?
+          first_break_index = first_break_index_one
+        end
+        unless first_break_index.nil? || allow_multi_paragraph_hint
```
No reason to have OR here if `allow_multi_paragraph_hint` is always false

klaasvandeweerdt replied:
The multiline logic was copied from the GraphQL connector. There is no need to support multi `get_allow_multi_paragraph_hint`. I removed all code for `allow_multi_paragraph_hint`. 
It will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+          first_break_index = first_break_index_one
+        end
+        unless first_break_index.nil? || allow_multi_paragraph_hint
+          description = description[0..first_break_index - 1]
```
probably you can convert to `description[0...first_break_index]`

klaasvandeweerdt replied:
The multiline logic was copied from the GraphQL connector. There is no need to support multi `get_allow_multi_paragraph_hint`. I removed all code for `allow_multi_paragraph_hint`. 
It will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2827 @@
+          description = description.gsub(/(.*)\W$/, '\1')
+          documentation_href = nil
+          if documentation_href.blank?
+            documentation_href = nil # call('get_documentation_href', connection)
```
Need to uncomment?

klaasvandeweerdt replied:
The multiline logic was copied from the GraphQL connector. There is no need to support multi `get_allow_multi_paragraph_hint`. I removed all code for `allow_multi_paragraph_hint`. 
It will be in the next commit.

den-sergeyev-workato commented:
@klaasvandeweerdt I may add more comments as I haven't finished reviewing everything.

den-sergeyev-workato commented:
@klaasvandeweerdt Is it possible to either fix TODOs and all commented lines of code in first version?

klaasvandeweerdt commented:
@den-sergeyev-workato, thanks. Let me start with the description/comments and highlights.
I will go through the TODOs and commented lines as well, most were copied of from the GraphQL connector, so it will take some time.

Commit:
Added review recommendations

klaasvandeweerdt commented:
@den-sergeyev-workato, I already implemented a part of the recommendation and comments, marked them as resolved, and push the latest version.

klaasvandeweerdt commented:
@den-sergeyev-workato, I had an issue when executing `Custom action`, the `operationName` value is not defined in the payload when the query contains multiple operations. I used the GraphQL connector code as the base for this connector. This issue also exists in the GraphQL connector.

Commit:
Added review recommendations

Commit:
Added review recommendations #3

klaasvandeweerdt commented:
@den-sergeyev-workato, implemented all recommendations and comments, the unresolved conversations are answers to your questions. can be resolved from my end unless there is still something that needs to be done.

Commit:
Added additional specs for custom operation with operation name

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+      field
+    end,
+
+    format_field_hint: lambda do |connection, description|
```
Unused connection parameter.

klaasvandeweerdt replied:
Removed, will be in the next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+    build_action_input_fields: lambda do |connection, input, operation_type|
+      problems = []
+      report_problem = ->(msg) { problems << msg }
+      error("#{problems.join(', ')}, ") unless problems.empty?
```
Weird error message, why do we want to append ", " in the end?

klaasvandeweerdt replied:
In the debugger, it appends line xxx after it. It looks nicer with a comma.
Removed the comma, will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+    build_action_input_fields: lambda do |connection, input, operation_type|
+      problems = []
+      report_problem = ->(msg) { problems << msg }
+      error("#{problems.join(', ')}, ") unless problems.empty?
```
Probably I'm missing something but how it can have something right after initialization here? Does it imply another behavior or this is just copy-paste from another place?

klaasvandeweerdt replied:
Updated, will be in next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+
+    does_require_argument: lambda do |field|
+      # get all arguments
+      args = field['args']
```
NIT: 
```
     # get all arguments
      args = field['args']

      # find a required argument
      required_argument = args&.find do |arg|
        call('is_non_null', arg['type'])
      end

      # return true if required argument is found
      required_argument.present?
```

Could be simplified to
```
  field['args']&.any? do |arg|
   call('is_non_null', arg['type'])
  end
```

klaasvandeweerdt replied:
Updated, will be in next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+          delimiter: ','
+        }
+
+        selected_field_names = input&.[]('primitive_fields')&.split(',')
```
Do we expect `input` is nil here?

klaasvandeweerdt replied:
Can be `nil`

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+        unless selected_field_names.present?
+          # default to non-required fields
+          selected_field_names = primitive_fields.map do |sub_field|
+            next if call('does_require_argument', sub_field)
```
do we need to `compact` result if we want to get rid of `nil`?

klaasvandeweerdt replied:
Updated. will be in next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+            connection,
+            report_problem,
+            output_field,
+            input&.[](field_name)
```
Do we need nil safe here?

klaasvandeweerdt replied:
Can be `nil`

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+            connection,
+            report_problem,
+            nested_field,
+            input&.[](input_name)
```
Do we need safe navigation here?

klaasvandeweerdt replied:
yes can be `nil`

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+      selected_fragment_types = input&.[]('expected_fragment_types')&.split(',')
+      selected_fragment_types&.each do |fragment_type_name|
+        # get schema type for selected object
+        schema_type = possible_types.find do |type|
```
What if possible type `nil` here?  Based on this:
```
possible_types = schema_type['possibleTypes']&.map do |possible_type|
        call(
          'get_schema_type',
          connection,
          report_problem,
          possible_type['kind'],
          possible_type['name']
        )
      end&.compact

klaasvandeweerdt replied:
Works fine. possible_types  is only used when present

den-sergeyev-workato replied:
Moved it into `possible_types.present?`, will be in the next commit.
> It is not inside `possible_types.present?` branch. I don't understand why it cannot fail with undefined method on nil?

klaasvandeweerdt replied:
Moved it into possible_types.present?, will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+          connection,
+          report_problem,
+          schema_type,
+          input[fragment_field[:name]]
```
Is it possible `input` is nil here? you treat as possible case above:
```
  selected_fragment_types = input&.[]('expected_fragment_types')&.split(',')
```

klaasvandeweerdt replied:
Yes it is.

den-sergeyev-workato replied:
so you need safe navigation here?

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+          type = sub_field['type']
+          next unless call('is_primitive_type', type)
+
+          # exclude from defaults if user input requred
```
Typo: `required`

klaasvandeweerdt replied:
Updated, will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+      primitive_fields = sub_fields.select do |sub_field|
+        type = sub_field['type']
+        call('is_primitive_type', type)
+      end.compact
```
What case does this `compact` handles? Select returns sub_fields and if any of them expected to be nil then `sub_field['type']` must fail.

klaasvandeweerdt replied:
Updated, will be in the next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+          next if call('does_require_argument', sub_field)
+
+          true
+        end.compact
```
Probably useless compact after select unless `sub_field` may be nil?

klaasvandeweerdt replied:
Updated will be in the next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+      end
+
+      # get output field names
+      primitive_field_names = input&.[]('primitive_fields')&.split(',')
```
is it possible to have nil for input?

klaasvandeweerdt replied:
Yes it can be `nil`

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+      properties = []
+      selected_field_names = (primitive_field_names || []) +
+                             (nested_field_names || [])
+      selected_field_names&.each do |field_name|
```
`selected_field_names` is never `nil`

klaasvandeweerdt replied:
Updated, will be in the next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+          field_name
+        )
+        properties << field
+      end&.compact
```
you don't save result or call in place method(`compact!`), how it is expected to work?

klaasvandeweerdt replied:
Updated, will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+        )
+        fragment_field[:properties]&.each do |field|
+          # check if field is already in properties
+          next if properties.find do |p|
```
NIT: too overcomplicated
```
 # check if field is already in properties
 next if properties.find do |p|
            p[:name] == field[:name]
          end.present?

          properties << field
```
vs
```
 properties << field if properties.none? { |p|  p[:name] == field[:name] }

```

klaasvandeweerdt replied:
Updated, will be in the next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+      field
+    end,
+
+    apply_field_default_value: lambda do |connection, report_problem, field, default_value|
```
`connection` parameter is not used.

klaasvandeweerdt replied:
Removed, will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+                            default_value.to_s
+                          else
+                            report_problem&.call(
+                              "Enexpected field type: #{field[:type]}"
```
Typo: 'Unexpected'

klaasvandeweerdt replied:
Updated, will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+                            nil
+                          else
+                            report_problem&.call(
+                              "Enexpected field type: #{field[:type]}"
```
Typo: 'Unexpected'

klaasvandeweerdt replied:
Updated, will be in the next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+        default_sentence = "Defaults to #{formatted_value}" if formatted_value.present?
+        if hint.present?
+          hint = "#{hint}." unless hint.ends_with?('.')
+          hint = "#{hint} #{default_sentence}"
```
Should we append empty default sentence if `formatted_value.present? == false`

klaasvandeweerdt replied:
Updated, will be in next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+      # treat blank string for optional fields as nil
+      value = nil if value.is_a?(String) && optional && value == ''
+
+      unless value.nil? && optional
```
NIT: I would recommend to convert to if condition, otherwise it is hard to follow logic.
So it executes "either value present or it is not optional"?

klaasvandeweerdt replied:
Updated, will be in next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+           (type == 'string' && control_type != 'select')
+          value = "\"#{value}\""
+        elsif type == 'array'
+          value = [value] unless value.is_a?(Array)
```
[Array.wrap?](https://docs.workato.com/developing-connectors/sdk/sdk-reference/ruby_methods.html#wrap)

klaasvandeweerdt replied:
Updated, will be in the next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+          fragment_type_names&.each do |fragment_type_name|
+            field_input = input&.[]("fragment_#{fragment_type_name}")
+            input_schema = eis&.find { |field| field['name'] == "fragment_#{fragment_type_name}" }&.[]('properties')
+            output_schema = eos
```
Unused output_schema var

klaasvandeweerdt replied:
Updated will be in the next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+          fields << '__typename' if fragment_type_names.present?
+        end
+      end
+      "{#{fields&.join(' ')}}" if fields.present?
```
`fields` are always present no need `&`

klaasvandeweerdt replied:
Updated, will be in the next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+    end,
+
+    action_execute: lambda do |connection, input, eis, eos, operation_type|
+      field_name = input&.[]('object')
```
input is alway present here?

klaasvandeweerdt replied:
Updated, will be in the next commit

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2815 @@
+          query,
+          nil,
+          nil,
+          input['account']
```
Do we need safe navigation as in 
```
field_name = input&.[]('object')
```

klaasvandeweerdt replied:
No, the account is a required input field for all GraphQL queries.

klaasvandeweerdt replied:
Updated, will be in the next commit

den-sergeyev-workato commented on fixtures/actions/confirm_webhook/input/webhook_verify.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "webhook_verify": "https://api.4me.com/v1"
+}
```
Can you, please, fix new line error in all files?

klaasvandeweerdt replied:
Updated, will be in the next commit

den-sergeyev-workato replied:
@klaasvandeweerdt doesn't seem like resolved, probably, you need to add this to your editor config so new files won't have this issue after every update?

klaasvandeweerdt replied:
@den-sergeyev-workato, no idea, let me check my editor, I added a new line at the end to all JSON files. Like this.
```
{\n
  "webhook_verify": "https://api.4me.com/v1"\n
}\n
```
A new commit in about 30 min.

den-sergeyev-workato commented on .gitignore:
```diff
@@ -7,6 +7,10 @@
 # rspec failure tracking
 .rspec_status
+
+/master.key
```
Won't  `master.key` pattern above cover this case as well?

klaasvandeweerdt replied:
Yes, updated, will be in the next commit.

den-sergeyev-workato commented on .gitignore:
```diff
@@ -7,6 +7,10 @@
 .rspec_status
+
+/master.key
+/tape_library.log
```
I believe `tape_library.log` is enough to ignore all such files in any folder

klaasvandeweerdt replied:
Yes, updated, will be in the next commit.

den-sergeyev-workato commented on spec/actions/query_spec.rb:
```diff
@@ -0,0 +1,33 @@
+
+    it 'contains totalCount greater than 0' do
+      # rubocop:disable Style/NumericPredicate
+      expect(output[:totalCount]).to be > 0
```
`be_positive` ?

klaasvandeweerdt replied:
Updated, will be in the next commit

den-sergeyev-workato commented on spec/methods/apply_field_default_value_spec.rb:
```diff
@@ -0,0 +1,75 @@
+
+    it 'adds default value to the field' do
+      apply_field_default_value
+      expect(field[:default].to_s).to eq('')
```
`be_empty`

klaasvandeweerdt replied:
Updated to `be_blank`, will be in the next commit

den-sergeyev-workato commented on spec/methods/apply_field_description_spec.rb:
```diff
@@ -0,0 +1,32 @@
+
+    it 'adds a hint to the field' do
+      apply_field_description
+      expect(field[:hint]).to eq('foo &#x2022; test')
```
Is it expected conversion from bar to &#x2022;?

klaasvandeweerdt replied:
Updated the code and specs, `* start with` should become `• start with`

den-sergeyev-workato commented on spec/methods/apply_field_description_spec.rb:
```diff
@@ -0,0 +1,32 @@
+  end
+
+  context 'does nothing when there is no description' do
+    let(:connection) { {} }
```
Yo can extract it on top level and override in context when necessary. Now you have duplicates.

klaasvandeweerdt replied:
Updated, will be in next commit

den-sergeyev-workato commented on spec/methods/apply_field_default_value_spec.rb:
```diff
@@ -0,0 +1,75 @@
+    end
+  end
+
+  context 'applies field default datetime value' do
```
How does it behave if date format is not parseble?

klaasvandeweerdt replied:
This is used to parse default value formats. We do not use default values for date/time, if we would use they will be ISO8601 compliant.

den-sergeyev-workato commented on spec/methods/parse_graphql_spec.rb:
```diff
@@ -0,0 +1,21 @@
+
+    it 'contains documents and fragments' do
+      expect(result[:documents]).to be_present
+      expect(result[:fragments]).to eq([])
```
fragments doesn't look like present?

klaasvandeweerdt replied:
Updated, will be in next commit

den-sergeyev-workato commented on spec/pick_lists/webhook_events_spec.rb:
```diff
@@ -0,0 +1,103 @@
+        ['Time entry create', 'time_entry.create'],
+        ['Time entry delete', 'time_entry.delete'],
+        ['Time entry update', 'time_entry.update'],
+        # ['Webhook verify','webhook.verify'],
```
Uncomment or remove?

klaasvandeweerdt replied:
Removed it, will be in the next commit

klaasvandeweerdt replied:
Removed, will be in the next commit.

den-sergeyev-workato commented on spec/spec_helper.rb:
```diff
@@ -22,17 +22,33 @@
 VCR.configure do |config|
   config.cassette_library_dir = 'tape_library'
   config.hook_into :webmock
+  # config.debug_logger = File.open('tape_library.log', 'w')
```
Remove or move to comment if needed for future?

klaasvandeweerdt replied:
Can be useful in the future, I moved it to the top and added a description.
Will be in the next commit.

klaasvandeweerdt replied:
Move it to the top and added a description on then to use it. Will be in the next commit.

den-sergeyev-workato commented on spec/spec_helper.rb:
```diff
@@ -22,17 +22,33 @@
+    decode_compressed_response: false,
     serialize_with: :encrypted,
-    match_requests_on: %i[uri headers_without_user_agent body]
+    # match_requests_on: %i[uri headers_without_user_agent body]
```
Remove?

klaasvandeweerdt replied:
Removed it, will be in the next commit.

den-sergeyev-workato commented on spec/triggers/new_event_spec.rb:
```diff
@@ -0,0 +1,267 @@
+    let(:input) { JSON.parse(File.read('fixtures/triggers/input/webhook.json')) }
+
+    it 'person.create' do
+      # TODO
```
Add body?

klaasvandeweerdt replied:
Removed it, as there is no body. Will be in the next commit.

den-sergeyev-workato commented on spec/triggers/new_event_spec.rb:
```diff
@@ -0,0 +1,267 @@
+    subject(:output) { trigger.dedup(record) }
+    let(:record) { { 'account_id' => 'wdc', 'event' => 'automation_rule', 'object_id' => 'abc' } }
+    it {
+      expect(output).to include('wdc_automation_rule_abc_')
```
Doesn't it expect timestamp as suffix?

klaasvandeweerdt replied:
Updated, will be in next commit

den-sergeyev-workato commented:
@klaasvandeweerdt Can you, please, elaborate on this?
> @den-sergeyev-workato, I had an issue when executing Custom action, the operationName value is not defined in the payload when the query contains multiple operations. I used the GraphQL connector code as the base for this connector. This issue also exists in the GraphQL connector.

den-sergeyev-workato commented:
@klaasvandeweerdt I've completed review and in general it looks good, there are several comments that are related to the last part I reviewed this week.

klaasvandeweerdt commented:
@den-sergeyev-workato, thanks. I will look at those later today. About the issue in the universal GraphQL connector.
When using the `Custom action` action I can execute any GraphQL query.

For example
```
query Query1($id: ID!) {
  node(id: $id) {
    ... on Person {
      id
      name
      primaryEmail
    }
  }
}
query Query2($id: ID!, $test: String) {
  node(id: $id) {
    ... on Person {
      id
      name
      primaryEmail
    }
  }
}
```
When I want to execute it I need to send an operation name, either name `Query1` or `Query2`.
The connector shows me a dropdown from which I can select one of the two defined operation names, based on the selection it shows me the input fields.

When I execute the query it returns the following error, `An operation name is required at line: 1154`
The payload sent to the GraphQL endpoint is the query and the variables.
```
{
  "query":"query Query1 ($id: ID!) { node(id: $id) { ... on Person { id name primaryEmail } } }\nquery Query2 ($id: ID!) { node(id: $id) { ... on Person { id name primaryEmail } } }",
  "variables":{"id":"abc"}
}
```
because there are two queries it is also required to specify the operation name. The query should be like this
```
{
  "query":"query Query1 ($id: ID!) { node(id: $id) { ... on Person { id name primaryEmail } } }\nquery Query2 ($id: ID!) { node(id: $id) { ... on Person { id name primaryEmail } } }",
  "operationName":"Query1",
  "variables":{"id":"abc"}
}
```
The selected operation name input value is not used when executing the query, which causes the error.
I already implemented it in the 4me connector and added a spec for it.

Commit:
Added review recommendations #4

klaasvandeweerdt commented:
@den-sergeyev-workato,
Implemented/answered all recommendations.
For the summarized input and output, do you have an example where it is generated dynamically? It is the same challenge as the samples, the input and output can be any object with any selection of fields and connections.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2787 @@
+            request.headers('x-4me-Account': connection['account'])
+            payload = {
+              grant_type: 'client_credentials'
+            }.compact
```
Should we compact after merge credentials? Otherwise I don't see how it can be useful here?

klaasvandeweerdt replied:
No, the merge is not needed either. This works as well.
```
payload = {
  grant_type: 'client_credentials',
  client_id: connection['client_id'],
  client_secret: connection['client_secret']
}
```
Will be in the next commit.

Commit:
Updated new_event_spec

den-sergeyev-workato commented on fixtures/actions/mutation/input/person_update.json:
```diff
@@ -0,0 +1,15 @@
+      "primitive_fields": "id"
+    }
+  }
+  
```
Several new lines are still needed

klaasvandeweerdt replied:
@den-sergeyev-workato, what do you mean by new lines?

den-sergeyev-workato replied:
<img width="800" alt="Screen Shot 2023-03-10 at 10 01 02 PM" src="https://user-images.githubusercontent.com/95912036/224427444-f75fcd8a-7897-4178-9cfd-90c78cbdbb79.png">

You have these formatting errors.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2787 @@
+               end
+
+        webhook_identifiers = input['webhook_identifier']
+        webhook_identifiers = [webhook_identifiers] unless webhook_identifiers.is_a?(Array)
```
Array.wrap ?

klaasvandeweerdt replied:
Updated, will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2787 @@
+        connection,
+        webhook_subscribe_output|
+
+        data = if input['webhook_policy'].blank?
```
Can you, please, explain why it can be blank and when it is used?

klaasvandeweerdt replied:
The code is used for a shared connector, this way our clients can test it as well. In that version JWT is optional.
I will update it and will be in the next commit.

klaasvandeweerdt replied:
I also added an additional check `if payload&.[]('jwt')`, anything else should be ignored.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2787 @@
+          {
+            ngIf: 'input.webhook_policy == "HS256" || input.webhook_policy == "HS384" || ' \
+                  'input.webhook_policy == "HS512"',
+            name: 'webhook_secret_hs',
```
It is preferable to have full `hmac` insted of `hs`

klaasvandeweerdt replied:
Updated to `webhook_secret_hmac`, will be in the next commit.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2787 @@
+                  'input.webhook_policy == "RS512" || input.webhook_policy == "ES256" || ' \
+                  'input.webhook_policy == "ES256" || input.webhook_policy == "ES384" || ' \
+                  'input.webhook_policy == "ES512"',
+            name: 'webhook_secret',
```
It is not really secret, it is public key, probably name it public key then?

klaasvandeweerdt replied:
Updated to `webhook_public_key`, will be in the next commit.

den-sergeyev-workato commented:
@klaasvandeweerdt Indeed there is no dynamic summarize option, we will add it to our backlog, so this is not a requirement for this review anymore.
> Implemented/answered all recommendations.
For the summarized input and output, do you have an example where it is generated dynamically? It is the same challenge as the samples, the input and output can be any object with any selection of fields and connections.

den-sergeyev-workato commented:
Ok, so you want to back import this change to universal connector. I will discuss it with CS team.

> @den-sergeyev-workato, thanks. I will look at those later today. About the issue in the universal GraphQL connector. When using the `Custom action` action I can execute any GraphQL query.
> 
> For example
> 
> ```
> query Query1($id: ID!) {
>   node(id: $id) {
>     ... on Person {
>       id
>       name
>       primaryEmail
>     }
>   }
> }
> query Query2($id: ID!, $test: String) {
>   node(id: $id) {
>     ... on Person {
>       id
>       name
>       primaryEmail
>     }
>   }
> }
> ```
> 
> When I want to execute it I need to send an operation name, either name `Query1` or `Query2`. The connector shows me a dropdown from which I can select one of the two defined operation names, based on the selection it shows me the input fields.
> 
> When I execute the query it returns the following error, `An operation name is required at line: 1154` The payload sent to the GraphQL endpoint is the query and the variables.
> 
> ```
> {
>   "query":"query Query1 ($id: ID!) { node(id: $id) { ... on Person { id name primaryEmail } } }\nquery Query2 ($id: ID!) { node(id: $id) { ... on Person { id name primaryEmail } } }",
>   "variables":{"id":"abc"}
> }
> ```
> 
> because there are two queries it is also required to specify the operation name. The query should be like this
> 
> ```
> {
>   "query":"query Query1 ($id: ID!) { node(id: $id) { ... on Person { id name primaryEmail } } }\nquery Query2 ($id: ID!) { node(id: $id) { ... on Person { id name primaryEmail } } }",
>   "operationName":"Query1",
>   "variables":{"id":"abc"}
> }
> ```
> 
> The selected operation name input value is not used when executing the query, which causes the error. I already implemented it in the 4me connector and added a spec for it.

den-sergeyev-workato commented:
@klaasvandeweerdt tests are passing locally, I'm going to prepare PR with connector integration to see if it causes any additional comments.

klaasvandeweerdt commented:
It is a bug in the universal connector. Based on the specific use case, it is rather a low priority, but adding in the next release would be nice I may save some time and frustration for other people.
> Ok, so you want to back import this change to universal connector. I will discuss it with CS team.
> 
> > @den-sergeyev-workato, thanks. I will look at those later today. About the issue in the universal GraphQL connector. When using the `Custom action` action I can execute any GraphQL query.
> > For example
> > ```
> > query Query1($id: ID!) {
> >   node(id: $id) {
> >     ... on Person {
> >       id
> >       name
> >       primaryEmail
> >     }
> >   }
> > }
> > query Query2($id: ID!, $test: String) {
> >   node(id: $id) {
> >     ... on Person {
> >       id
> >       name
> >       primaryEmail
> >     }
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > When I want to execute it I need to send an operation name, either name `Query1` or `Query2`. The connector shows me a dropdown from which I can select one of the two defined operation names, based on the selection it shows me the input fields.
> > When I execute the query it returns the following error, `An operation name is required at line: 1154` The payload sent to the GraphQL endpoint is the query and the variables.
> > ```
> > {
> >   "query":"query Query1 ($id: ID!) { node(id: $id) { ... on Person { id name primaryEmail } } }\nquery Query2 ($id: ID!) { node(id: $id) { ... on Person { id name primaryEmail } } }",
> >   "variables":{"id":"abc"}
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > because there are two queries it is also required to specify the operation name. The query should be like this
> > ```
> > {
> >   "query":"query Query1 ($id: ID!) { node(id: $id) { ... on Person { id name primaryEmail } } }\nquery Query2 ($id: ID!) { node(id: $id) { ... on Person { id name primaryEmail } } }",
> >   "operationName":"Query1",
> >   "variables":{"id":"abc"}
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The selected operation name input value is not used when executing the query, which causes the error. I already implemented it in the 4me connector and added a spec for it.

klaasvandeweerdt commented:
@den-sergeyev-workato, ok, thanks. Is there a mailing list/subscription specifically for the Connector SDK? Would be a good way to stay up to date, and update the Connector if there would be features relevant to us.

> @klaasvandeweerdt Indeed there is no dynamic summarize option, we will add it to our backlog, so this is not a requirement for this review anymore.
> 
> > Implemented/answered all recommendations.
> > For the summarized input and output, do you have an example where it is generated dynamically? It is the same challenge as the samples, the input and output can be any object with any selection of fields and connections.

Commit:
Added review recommendations #5

klaasvandeweerdt commented:
@den-sergeyev-workato, Implemented/answered all recommendations. I also added additional specs.

Commit:
Updated gems - Workato 1.3.3

Commit:
Added webhook notification spec

Commit:
Added display_priority to actions

den-sergeyev-workato commented:
@klaasvandeweerdt I've pulled latest tests and 4 of them are failing, can you, please, have a look?
```

rspec ./spec/triggers/new_event_spec.rb:20 # triggers/new_event webhook notification - using SHA-256 is present
rspec ./spec/triggers/new_event_spec.rb:24 # triggers/new_event webhook notification - using SHA-256 contains event, data and payload
rspec ./spec/triggers/new_event_spec.rb:30 # triggers/new_event webhook notification - using SHA-256 event equals webhook.verify
rspec ./spec/triggers/new_event_spec.rb:34 # triggers/new_event webhook notification - using SHA-256 payload and data start with https://

```

klaasvandeweerdt commented:
@den-sergeyev-workato, working locally for me, but I had to update the Workato gem to 1.3.3, jwt_decode wasn't working.

> @klaasvandeweerdt I've pulled latest tests and 4 of them are failing, can you, please, have a look?
> 
> ```
> 
> 
> 
> rspec ./spec/triggers/new_event_spec.rb:20 # triggers/new_event webhook notification - using SHA-256 is present
> 
> rspec ./spec/triggers/new_event_spec.rb:24 # triggers/new_event webhook notification - using SHA-256 contains event, data and payload
> 
> rspec ./spec/triggers/new_event_spec.rb:30 # triggers/new_event webhook notification - using SHA-256 event equals webhook.verify
> 
> rspec ./spec/triggers/new_event_spec.rb:34 # triggers/new_event webhook notification - using SHA-256 payload and data start with https://
> 
> 
> 
> ```

den-sergeyev-workato commented:
@klaasvandeweerdt, it is green for me too, probably I didn't use the latest gem for some reason.

Commit:
Updated fixtures json files

klaasvandeweerdt commented:
@den-sergeyev-workato, what is PR?
> @klaasvandeweerdt tests are passing locally, I'm going to prepare PR with connector integration to see if it causes any additional comments.

den-sergeyev-workato commented:
@klaasvandeweerdt After work completion here we introduce new platform connector through internal PR in our repository, it passes additional checks, reviews and testing to become part of Workato codebase. At the moment our internal PR is raised.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -1,37 +1,2789 @@
+          webhook_identifiers = Array.wrap(webhook_identifiers)
+
+          if webhook_identifiers.include?(data['webhook_nodeID']) &&
+             (data['event'] == input['event_selection'] || data['event'] == 'webhook.verify')
```
@klaasvandeweerdt 
Question raised during internal review: 
Is it possible to move this check to `webhook_key`/`webhook_keys` so it will be more efficient routing to specific webhooks, not all of them?

`webhook_keys` block does not support `workato.jwt_decode` but it supports regular methods on strings `decode_base64` and `split`. 

Do you think it is possible to parse payload for routing without signature verification. In webhook_notification you still need to make proper decode.

klaasvandeweerdt replied:
@den-sergeyev-workato 
Yes, I can make the changes.
One question, when I make the change then I also lose the functionality of specifying multiple webhook identifiers?
The result of the `webhook_key` should be an exact match with the value returned in `webhook_keys`. In the `webhook_notification` now use `webhook_identifiers.include?(data['webhook_nodeID'])`. The webhook input in formula mode allows me to input an array instead of a string.

klaasvandeweerdt replied:
@den-sergeyev-workato
I spend quite some time changing the webhook routing but got it working, will be in the next commit.

den-sergeyev-workato commented:
@klaasvandeweerdt we are going to release  `automatic webhook verification` requested by your team this week, are you going to update the connector regarding this? We are running tests with current version right now.

Can you, please, open the PR to main branch with the latest version so I can close this PR(opened by me).

klaasvandeweerdt commented:
@den-sergeyev-workato
I will remove the webhook verification actions and specs once the test is successful and then create a new PR.

> @klaasvandeweerdt we are going to release `automatic webhook verification` requested by your team this week, are you going to update the connector regarding this? We are running tests with current version right now.
> 
> Can you, please, open the PR to main branch with the latest version so I can close this PR(opened by me).

Commit:
Removed webhook confirmation action and updated webhook_keys to decode the incoming JWT request

klaasvandeweerdt commented:
@den-sergeyev-workato, can I create a new PR from the 4me branch when this PR is still open?

> @den-sergeyev-workato I will remove the webhook verification actions and specs once the test is successful and then create a new PR.
> 
> > @klaasvandeweerdt we are going to release `automatic webhook verification` requested by your team this week, are you going to update the connector regarding this? We are running tests with current version right now.
> > Can you, please, open the PR to main branch with the latest version so I can close this PR(opened by me).

den-sergeyev-workato commented:
> @den-sergeyev-workato, can I create a new PR from the 4me branch when this PR is still open?
> 
> > @den-sergeyev-workato I will remove the webhook verification actions and specs once the test is successful and then create a new PR.
> > > @klaasvandeweerdt we are going to release `automatic webhook verification` requested by your team this week, are you going to update the connector regarding this? We are running tests with current version right now.
> > > Can you, please, open the PR to main branch with the latest version so I can close this PR(opened by me).

@klaasvandeweerdt Sure, you can create another branch from `4me` and create new pull request.

klaasvandeweerdt commented:
> > @den-sergeyev-workato, can I create a new PR from the 4me branch when this PR is still open?
> > > @den-sergeyev-workato I will remove the webhook verification actions and specs once the test is successful and then create a new PR.
> > > > @klaasvandeweerdt we are going to release `automatic webhook verification` requested by your team this week, are you going to update the connector regarding this? We are running tests with current version right now.
> > > > Can you, please, open the PR to main branch with the latest version so I can close this PR(opened by me).
> 
> @klaasvandeweerdt Sure, you can create another branch from `4me` and create new pull request.

@den-sergeyev-workato, I created the new branch and pull request

den-sergeyev-workato commented:
Closed in favor of [Author PR](https://github.com/Workato-Partners/4me/pull/3)
